### PR TITLE
Qiskit Project RFC Process

### LINKED ISSUE #173: Proposals/RFCs: Communicate and collaborate on broad design changes
Currently, there is no forum for proposing large scale Qiskit changes to our developers and community members. Github issues are better suited to smaller changes and do not easily serve as a wide-reaching, persistent and amendable document. Consider this recent [Terra issue](https://github.com/Qiskit/qiskit-terra/issues/1713) which is providing an overview of the proposed Pulse API. It was necessary to create this as a Github issue to expose it to the community. This is not the proper medium since it is difficult to track changes to the document and have community members review specific sections.

Other opensource projects have created different solutions for the above problem and have created a proposal document repositories with clear submission and acceptance guidelines:
- Python: [PEPs](https://www.python.org/dev/peps/)
- Tensorflow:  [TF-RFCs](https://github.com/tensorflow/community/blob/master/governance/TF-RFCs.md)

### PR DESCRIPTION
**Pull request by taalexander**
### Summary
This PR introduces an RFC process to the Qiskit organization.

The purpose of a Qiskit Request for Comments (RFC) is to communicate and engage with the community in the development and direction of Qiskit. RFCs enable engineering and research stakeholders to communicate large design changes.

Closes #173.

### Details and comments
To implement an effective RFC process requires buy-in from the organization. Please leave your comments, opinions, and suggestions on what you think would lead to an effective RFC process here. We should receive input from the community, and key stakeholders of the [Qiskit meta-package](https://github.com/Qiskit/qiskit), [Terra](https://github.com/Qiskit/qiskit-terra), [Aer](https://github.com/Qiskit/qiskit-aer),  [Aqua](https://github.com/Qiskit/qiskit-aqua), [Ignis](https://github.com/Qiskit/qiskit-ignis), and the [provider](https://github.com/Qiskit/qiskit-ibmq-provider), prior to acceptance and merging. 

Examples of RFCs in other large open source projects:
- [ietf](https://www.rfc-editor.org/rfc-index.html)
- [peps](https://www.python.org/dev/peps/)
- [neps](https://numpy.org/neps/)
- [tensorflow rfc](https://github.com/tensorflow/community/blob/master/governance/TF-RFCs.md)
- [rust rfc](https://github.com/rust-lang/rfcs)

### Questions
- Do we wish RFC to be in the Qiskit meta-package, or should they be their own repository? The downside of going in the Qiskit meta-package is that they can be more difficult to find, and will be included in all downloads of the package.

### CODE REVIEW

**Comment by dongreenberg** [comment]
Perhaps this is for lack of understanding of the purpose of this repo, but why is this in the `Qiskit/qiskit` repo? Several others seem to put RfCs into their own repo, which makes them easier to find for newcomers. There is no dependency that I see between these and the documentation, and the issue and PR lists will be an unrelated mix of RfCs and documentation changes. I'm usually an advocate of merging repos together... but in this case I'm not sure.

**Comment by dongreenberg** [comment]
That said, I am hugely in favor of this and excited to see it happening. Thank you, Thomas!!!

**Comment by mtreinish** [comment]
I think doing it in qiskit/qiskit is probably the proper place for things that cover multiple repos. If a feature is isolated to a single repo we can use the same process isolated in that single repository.

File: rfcs/.gitignore [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# Byte-compiled / optimized / DLL files
```
Comment (mtreinish) [inline rfcs/.gitignore:1]: Is this file needed?

**Review by mtreinish — COMMENTED** [review]
So the plan here to land a commit in this directory for RFCs? I envisioned this being done in issues. The template would fit fine as a new issue type and we can easily have discussion there since the difference between PRs and issues in github is minimal except for merging code to the repo. We can easily track status of an RFC with a tag in issues (and also automate it easily using a bot if needed). If we feel that PRs are the way we want to do something like this than I think we should do this in a separate repo (definitely not in qiskit/qiskit). Merges to all the elements have a lot of post merge jobs which don't have anything to do with rfcs (which is part of why I think issues are a better way to handle them). Especially in qiskit/qiskit which publishes docs to qiskit.org on merges.

**Comment by mtreinish** [comment]
Oh, also another advantage of using a separate repo with the PR model is we can publish the rendered markdown on a github page like https://qiskit.github.io/rfcs (this doesn't actually exist, it's just an example) This is a model I've used in other projects before, see: http://specs.openstack.org/openstack/nova-specs/specs/train/index.html for example. (although rst/sphinx was used for that, but the specific tools don't matter).

I guess the tl;dr of what I'm saying is that I think we should either do RFCs as either issues on the existing repos or as PRs on a dedicate repo. I think I'm leaning towards issues right now (which is what I was assuming this was when I wrote the first comment), but either would work well.

**Comment by dongreenberg** [comment]
I am more in favor of PRs than issues because:

1. I think the list of historical RfCs is important context and should be easily browsed, and keeping them in a directory is the most transparent way to do this IMO. Old issues generally are hidden unless we intend to keep every RfC issue open to be browsable (which is not the expectation with issues and will be annoying for the documentation people). They'll be especially hard to find amid the many closed documentation issues.
2. Commenting on individual lines of an RfC is critical and this is much more easily done and managed in PRs than by replying to an issue
3. RfCs can take a while to write. Collaboration, saving progress, and branching are more naturally done in a PR than an issue.

**Comment by mtreinish** [comment]
> I am more in favor of PRs than issues because:
> 
>     1. I think the list of historical RfCs is important context and should be easily browsed, and keeping them in a directory is the most transparent way to do this IMO. Old issues generally are hidden unless we intend to keep every RfC issue open to be browsable (which is not the expectation with issues and will be annoying for the documentation people). They'll be especially hard to find amid the many closed documentation issues.
> 
>     2. Commenting on individual lines of an RfC is critical and this is much more easily done and managed in PRs than by replying to an issue
> 
>     3. RfCs can take a while to write. Collaboration, saving progress, and branching are more naturally done in a PR than an issue.

1 is actually pretty easy to solve with issues, besides just the generic rfc view using tags/labels. For example looking at https://github.com/Qiskit/qiskit-terra/issues?page=1&q=is%3Aissue+label%3Adocumentation we can also leverage additional tags to indicate state and just filter on those as well. But 2 and 3 are valid and good reasons for going with PRs. (and more than good enough imo). I was leaning towards issues because it's easier to link to code PRs and track things that way, and also to have RFCs live closer to the code. But that process question is also easily solvable with some automation, which is something straightforward enough to add to https://github.com/Qiskit/qiskit-bot.

If we get sufficient buy in on this approach from other team members I can work with @taalexander on creating a new repo for this, because overloading this repo to handle an rfc process I don't think is a good idea (for the reasons I mentioned earlier). But I think using this PR as a discussion point before adopting this model is a good idea.

**Comment by taalexander** [comment]
I think everyone makes good and valid points above. I believe,

> Commenting on individual lines of an RfC is critical and this is much more easily done and managed in PRs than by replying to an issue

and 

> RfCs can take a while to write. Collaboration, saving progress, and branching are more naturally done in a PR than an issue.

are critical points. It would also be nice to be able to publish them as you say @mtreinish.

Lets keep the discussion going and try and source more input from team members.

**Comment by nonhermitian** [comment]
Is there a tensorflow or rust example of an RFC that was done before they hit stable 1.0?

**Comment by mtreinish** [comment]
> Is there a tensorflow or rust example of an RFC that was done before they hit stable 1.0?

Rust started their rfc process in the lead up to 1.0. Their rfcs started in earlier 2014: https://github.com/rust-lang/rfcs/blob/master/text/0001-private-fields.md and https://github.com/rust-lang/rfcs/blob/master/text/0002-rfc-process.md but the 1.0 release wasn't until May 2015. It was used pretty heavily towards getting to rust 1.0.

As for tensorflow it looks like they adopted their community RFC workflow about a year after their 1.0 release.

**Comment by taalexander** [comment]
I believe we have waited sufficient time for comments to come in. Given the relatively positive feedback here, @nonhermitian, do you think we might move forward by assigning an initial review committee to shepherd this through acceptance?

**Comment by nonhermitian** [comment]
Yes, that sounds like a good idea.

**Comment by taalexander** [comment]
Ok, one way to do this would be to assign a member from each major Qiskit project?

**Comment by taalexander** [comment]
Another example of the need for RFC as design documents is this [recent PR](https://github.com/Qiskit/qiskit-aqua/pull/740/files#diff-3a52961c19458a29a93e2798988b8e3c) by @dongreenberg. I would argue that since changes will not always be local to a single repository, there should be a global repository for `Qiskit` wide RFCs

File: rfcs/.gitignore (hunk 1 above)
Comment (1ucian0) [reply rfcs/.gitignore:1]: I dont think is needed.

**Comment by mtreinish** [comment]
I've created the rfcs repo to implement this: https://github.com/Qiskit/rfcs I'm going to close this now since it's been implemented over there. We can iterate on it over there now. Thanks for starting this process @taalexander
